cpp/velox/shuffle/VeloxShuffleWriter.cc

```diff
@@ -1134,50 +1152,55 @@ arrow::Status VeloxShuffleWriter::splitFixedWidthValueBuffer(const velox::RowVec
     return arrow::Status::OK();
   }
 
+  inline bool VeloxShuffleWriter::needRealloc(uint32_t partitionId, uint64_t newSize) {
+    auto currentBufferSize = partition2BufferSize_[partitionId];
+    return newSize > (1 + options_.buffer_realloc_threshold) * currentBufferSize ||
+        newSize < (1 - options_.buffer_realloc_threshold) * currentBufferSize;
+  }
+
   uint32_t VeloxShuffleWriter::calculatePartitionBufferSize(const velox::RowVector& rv) {
-    uint32_t sizePerRow = 0;
+    uint32_t bytesPerRow = 0;
     auto numRows = rv.size();
-    for (size_t i = fixedWidthColumnCount_; i < simpleColumnIndices_.size(); ++i) {
-      auto index = i - fixedWidthColumnCount_;
-      if (binaryArrayEmpiricalSize_[index] == 0) {
-        auto column = rv.childAt(simpleColumnIndices_[i]);
-        auto stringViewColumn = column->asFlatVector<velox::StringView>();
-        assert(stringViewColumn);
-
-        // accumulate length
-        uint64_t length = stringViewColumn->values()->size();
-        for (auto& buffer : stringViewColumn->stringBuffers()) {
-          length += buffer->size();
-        }
-
-        binaryArrayEmpiricalSize_[index] = length % numRows == 0 ? length / numRows : length / numRows + 1;
+    // Calculate average size bytes (bytes per row) for each binary array.
+    std::vector<uint64_t> binaryArrayAvgBytesPerRow(binaryColumnIndices_.size());
+    for (size_t i = 0; i < binaryColumnIndices_.size(); ++i) {
+      auto column = rv.childAt(binaryColumnIndices_[i]);
+      auto stringViewColumn = column->asFlatVector<velox::StringView>();
+      assert(stringViewColumn);
+
+      uint64_t binarySizeBytes = stringViewColumn->values()->size();
+      for (auto& buffer : stringViewColumn->stringBuffers()) {
```

Comment:
@FelixYBW We also iterates over stringBuffers, not each StringView.

Reply:
It's each StringView here